crates/edr_napi/src/provider.rs

```diff
@@ -18,13 +22,19 @@ pub struct Provider {
 impl Provider {
     #[doc = "Constructs a new provider with the provided configuration."]
     #[napi(ts_return_type = "Promise<Provider>")]
-    pub fn with_config(env: Env, config: ProviderConfig) -> napi::Result<JsObject> {
+    pub fn with_config(
+        env: Env,
+        config: ProviderConfig,
+        #[napi(ts_arg_type = "(message: Buffer) => void")] console_log_callback: JsFunction,
+    ) -> napi::Result<JsObject> {
         let config = edr_provider::ProviderConfig::try_from(config)?;
         let runtime = runtime::Handle::current();
 
+        let callbacks = Arc::new(InspectorCallback::new(&env, console_log_callback)?);
```

Comment:
Does this need to be an `Arc`? Or could you use a `Box<dyn ..>` and pass in a `&dyn` to the evm call?

Reply:
Good question! It's difficult to make what you proposed work, because then we would have to make `EvmInspector` take a reference to the callbacks, but `edr_evm::mine_block` expects a `Box<dyn SyncInspector<BlockchainErrorT, StateErrorT>>`, because `edr_evm::InspectorContainer` expects owned values. This means that there is a `static` bound on the lifetime of the reference passed to `EvmInspector` which we can only satisfy with owned values. 

To get around this, we'd have to change `edr_evm::InspectorContainer` to take `&mut dyn SyncInspector` which I tried, but it propagates `&mut` all over the place and leads to really unwieldy interfaces.

Reply:
I was curious to see whether I could pull it off. I added two commits to the PR to accomplish it:
- https://github.com/NomicFoundation/hardhat/pull/4659/commits/81d0b161a63ca498129993a724164ef6f538fa1f
- https://github.com/NomicFoundation/hardhat/pull/4659/commits/8694c00948d19f57133208520c0e46c9a574fcf1

The trickiest part was an issue with the borrow checker that erroneously complained about the lifetimes extending outside of the scope of the function blocks in `spawn_blocking`. I was able to work around that with `if let Some(..)`

Reply:
Nice, thanks! It's nice to see this work without having to make methods of `ProviderData``&mut`. Now that I think about it, when I tried this two weeks ago, I probably didn't add the `EvmInspector` wrapper yet, and had `ProviderData` take a `SyncInspector` argument instead of a `SyncInspectorCallbacks` which caused the `&mut` propagation in `ProviderData`.